dtm0/recovery.c

```diff
@@ -1686,11 +1726,40 @@ static void restore(struct m0_fom *fom,
 
 	do {
 		M0_SET0(&record);
-		rc = recovery_machine_log_iter_next(rf->rf_m, &rf->rf_log_iter,
-						    &rf->rf_tgt_svc, NULL,
-						    &record);
-		/* Any value except zero means that we should stop recovery. */
-		F(next) = rc == 0;
+		if (F(last_dtx_met)) {
+			/*
+			 * Last iteration reached the RECOVERING mark in the
+			 * log, don't need to send REDOs past this mark.
+			 */
+			F(next) = false;
+		} else {
+			do {
+				rc = recovery_machine_log_iter_next(
+						rf->rf_m, &rf->rf_log_iter,
+						&record);
+				/*
+				 * Any value except zero means that we should
+				 * stop recovery.
+				 *
+				 * XXX: handle cases when rc not in (0,
+				 * -ENOENT).
+				 */
+				F(next) = (rc == 0);
+				if (!F(next))
+					break;
+				rc = recovery_machine_dtxid_cmp(
+						rf->rf_m,
+						&record.dlr_txd.dtd_id,
+						&F(last_dtx_id));
+				F(last_dtx_met) = (rc == 0);
+				if (participated(&record, &rf->rf_tgt_svc))
```

Comment:
I have a question here.
Suppose there are 5 CAS services in the Motr system.
CAS0, CAS1, CAS2, CAS3, CAS4.
and we have 1+4 layout.
If CAS0 is crashed and restarted, then CAS[1-4] will all send REDO messages to CAS0.
But these messages are for the SAME DIX operation (duplicated to different CAS services).
The first REDO message arrived at CAS0 will be replayed successfully.
But the second/third/fourth REDO messages will fail.
Are there any de-dup mechanism to handle this?